File: apps/android/app/src/main/java/com/exponential/app/data/api/NotificationsApi.kt
package com.exponential.app.data.api

import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

// Mirrors apps/web/src/lib/trpc/notifications.ts — inbox mark-read.
@Serializable
private data class NotificationIdInput(@SerialName("id") val id: String)

@Serializable
private data class NotificationIssueInput(@SerialName("issueId") val issueId: String)

@Serializable
private object NotificationsEmptyInput

@Singleton
class NotificationsApi @Inject constructor(private val trpc: TrpcClient) {

    suspend fun markRead(accountId: String, id: String) {
        trpc.mutationUnit(
            accountId,
            path = "notifications.markRead",
            input = NotificationIdInput(id),
            inputSerializer = NotificationIdInput.serializer(),
        )
    }

    /** Opening an issue clears its inbox entries (EXP-92) — the read-on-open
     * safety net for push taps and app links that skip the inbox. */
    suspend fun markReadByIssue(accountId: String, issueId: String) {
        trpc.mutationUnit(
            accountId,
            path = "notifications.markReadByIssue",
            input = NotificationIssueInput(issueId),
            inputSerializer = NotificationIssueInput.serializer(),
        )
    }

    suspend fun markAllRead(accountId: String) {
        trpc.mutationUnit(
            accountId,
            path = "notifications.markAllRead",
            input = NotificationsEmptyInput,
            inputSerializer = NotificationsEmptyInput.serializer(),
        )
    }
}

File: apps/ios/ExpCore/Sources/API/NotificationsApi.swift
import Foundation

// Mirrors apps/web/src/lib/trpc/notifications.ts — inbox mark-read.
private struct NotificationIdInput: Encodable {
    let id: String
}

private struct NotificationIssueInput: Encodable {
    let issueId: String
}

private struct EmptyInput: Encodable {}

private struct EmptyResult: Decodable {}

public final class NotificationsApi: Sendable {
    private let trpc: TrpcClient

    public init(trpc: TrpcClient) {
        self.trpc = trpc
    }

    public func markRead(accountId: String, id: String) async throws {
        let _: EmptyResult = try await trpc.mutation(
            accountId: accountId,
            path: "notifications.markRead",
            input: NotificationIdInput(id: id)
        )
    }

    /// Opening an issue clears its inbox entries (EXP-92) — the read-on-open
    /// safety net for push taps and universal links that skip the inbox.
    public func markReadByIssue(accountId: String, issueId: String) async throws {
        let _: EmptyResult = try await trpc.mutation(
            accountId: accountId,
            path: "notifications.markReadByIssue",
            input: NotificationIssueInput(issueId: issueId)
        )
    }

    public func markAllRead(accountId: String) async throws {
        let _: EmptyResult = try await trpc.mutation(
            accountId: accountId,
            path: "notifications.markAllRead",
            input: EmptyInput()
        )
    }
}
